### PR TITLE
[addons] improve downgrading/version changing support

### PR DESCRIPTION
**Pull request by tamland**
This adds various client changes needed to support having multiple versions in the same repository and for installing specific versions from specific repositories.

On the gui side it changes the 'update' functionality in the info dialog to show a list of available versions and repositories which can be installed. Note that requires server side support. Currently there is only one version kept in the official repository.

### CODE REVIEW

File: xbmc/addons/GUIDialogAddonInfo.cpp [hunk 1]
```diff
@@ -197,8 +198,57 @@ void CGUIDialogAddonInfo::UpdateControls()
 
 void CGUIDialogAddonInfo::OnUpdate()
 {
-  CAddonInstaller::GetInstance().InstallOrUpdate(m_addon->ID());
   Close();
+
+  if (!m_localAddon)
+    return;
+
+  auto* dialog = static_cast<CGUIDialogSelect*>(g_windowManager.GetWindow(WINDOW_DIALOG_SELECT));
```
Comment (mkortstiege) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: I'd move this block to l235.

File: xbmc/addons/GUIDialogAddonInfo.cpp [hunk 2]
```diff
@@ -197,8 +198,57 @@ void CGUIDialogAddonInfo::UpdateControls()
...
+  dialog->Reset();
+  dialog->SetHeading(CVariant{21338});
+  dialog->SetUseDetails(true);
+
+  CAddonDatabase database;
+  database.Open();
```
Comment (mkortstiege) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: Please add a check like we do in the other functions.

File: xbmc/addons/AddonDatabase.cpp [hunk 3]
```diff
@@ -201,6 +201,67 @@ AddonVersion CAddonDatabase::GetAddonVersion(const std::string &id)
...
+  {
+    if (NULL == m_pDB.get()) return false;
+    if (NULL == m_pDS.get()) return false;
+
+    std::string sql = PrepareSQL(
+        "SELECT addon.version, repo.addonID AS repoID FROM addon "
+        "JOIN addonlinkrepo ON addonlinkrepo.idAddon=addon.id "
+        "JOIN repo ON repo.id=addonlinkrepo.idRepo "
+        "WHERE NOT EXISTS (SELECT * FROM  disabled WHERE disabled.addonID=repoID) "
+        "AND NOT EXISTS (SELECT * FROM  broken WHERE broken.addonID=addon.addonID) "
```
Comment (Razzeee) [inline xbmc/addons/AddonDatabase.cpp]: This row and the one above should probably be changed to ```SELECT TOP 1 1 FROM ...```
Saves some overhead.

File: xbmc/addons/GUIDialogAddonInfo.cpp (hunk 1 above)
Comment (tamland) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: Then it won't compile:)

File: xbmc/addons/AddonDatabase.cpp (hunk 3 above)
Comment (tamland) [inline xbmc/addons/AddonDatabase.cpp]: I'm no sql expert but first result on stackoverflow told me that was a nonsensical micro optimization..

File: xbmc/addons/GUIDialogAddonInfo.cpp (hunk 1 above)
Comment (mkortstiege) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: What?

**Comment by mkortstiege** [comment]
You might want to remove the `DialogSelect` changes a you've PR'd that one separately.

File: xbmc/addons/AddonDatabase.cpp (hunk 3 above)
Comment (Razzeee) [inline xbmc/addons/AddonDatabase.cpp]: Yep, it's micro tuning absolutely. Just ignore it :)

File: xbmc/addons/GUIDialogAddonInfo.cpp (hunk 1 above)
Comment (tamland) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: It's referenced before that, but I'm assuming you meant 'move it a bit further down'

Comment (mkortstiege) [inline xbmc/addons/GUIDialogAddonInfo.cpp]: Yes. Makes no sense to reset and semi-setup the dialog in case we're going to early return.

**Comment by tamland** [comment]
jenkins build this please

**Comment by tamland** [comment]
@mkortstiege good to go?

**Comment by mkortstiege** [comment]
Looks good. jenkins build and merge.

**Comment by MartijnKaijser** [comment]
I wonder why the "Upgrade" button is enabled if i just installed the latest version from repo?
Opening it just gives me the same one i already have and clicking it just re-install the same. Doesn't really make sense to allow that.

**Comment by tamland** [comment]
It doesn't strictly do "upgrading" any more but opens a window where you can manage/see all available versions (i plan to put rollback in there as well). It doesn't make any sense to disable that in some special cases. It will show an empty list if it's in no repos, it shows one item highlighted you have the latest version. The reason it reinstalls is because we don't track what is installed and where they come from, so there's no way to know it's the same one; other than comparing the version.

**Pull request merged**